### DISCUSSION TITLE
能否从dapp设计者的角度来串讲一下zk知识

### DISCUSSION THREAD

**Discussion by eiTT-01** [zkp-co-learning/zkp-co-learn#55] [Q&A（提问题在隔壁~）] [answered]
# 背景

课程主要围绕的是算法和数学原理的基础知识，这虽然是我们需要花时间了解的，但难以与实际问题联系起来，有种管中窥豹的感觉。

# 目标

希望可以通过一个实际问题或项目出发，来讲讲如何利用零知识证明来设计并构建一个应用，从而让我们对零知识证明有一个更全局的认识。

# 建议

可以从以下几个方面展开讲下，尽量先简化其中的数学原理

1. ZK-dapp 的用户交互流程，比如

<img width="470" alt="image" src="https://user-images.githubusercontent.com/31120840/224234507-706df8d0-5268-4f77-9717-75de48c73ba4.png">

2. 项目开发流程， 比如

      a. 编写零知识电路
      b. 生成验证用的智能合约
      c. 编写业务逻辑的智能合约，并集成上述验证合约
      d. 部署智能合约
      e. 本地客户端生成证明，并提交链上验证

3. 如何进行需求分析，并转换成电路
4. 如何设计约束，如何找出其中的问题并完善
5. 在整个过程中（可能主要就是电路编译、生成证明以及验证的过程），每个阶段是如何跟课程中知识关联

**Comment by Mobius3-3** [comment, marked as answer]
相关内容的公开资料很丰富，我post一些希望对你有帮助：
1. 从需求出发如何根据不同zk技术的tradeoffs进行方案设计:
（ZK Application Design Patterns | Devcon Bogotá）
https://www.youtube.com/watch?v=-PUKinjbLR8&list=PLaM7G4Llrb7x5JgkfCN9kf7IB6QtomyCu&index=24
2. 从隐私保护的需求出发，如何把zkSnark fit into现有应用：
https://vitalik.ca/general/2022/06/15/using_snarks.html
3. circom+snarkjs的电路编译、生成证明以及验证的过程：
https://github.com/iden3/snarkjs
4. 具体应用的交互建议看一些zk dapp的源码以及在本地跑一跑客户端，这样会有更直观的理解：
https://github.com/vplasencia/zkSudoku

**Reply by eiTT-01** [reply]
Thx
